lxd/tokens.go

```diff
@@ -66,6 +103,40 @@ func autoRemoveExpiredTokens(ctx context.Context, s *state.State) {
 	logger.Debug("Done removing expired tokens")
 }
 
+func getExpiredPendingIdentities(ctx context.Context, s *state.State) ([]cluster.Identity, error) {
+	var pendingTLSIdentities []cluster.Identity
+	err := s.DB.Cluster.Transaction(ctx, func(ctx context.Context, tx *db.ClusterTx) error {
+		var err error
+		dbPendingIdentityType := cluster.IdentityType(api.IdentityTypeCertificateClientPending)
+		pendingTLSIdentities, err = cluster.GetIdentitys(ctx, tx.Tx(), cluster.IdentityFilter{Type: &dbPendingIdentityType})
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	expiredPendingTLSIdentities := make([]cluster.Identity, 0, len(pendingTLSIdentities))
+	for _, pendingTLSIdentity := range pendingTLSIdentities {
+		metadata, err := pendingTLSIdentity.PendingTLSMetadata()
+		if err != nil {
```

Comment:
Can this be written without the duplicated append doing something like this (i.e in this case the happy path is an expired token)?

```go
metadata, err := pendingTLSIdentity.PendingTLSMetadata()

if metadata != nil &&  (metadata.Expiry.IsZero() || metadata.Expiry.After(time.Now())) {
    continue // Token has not expired.
} 

if err != {
    // If we fail to unmarshal it, it's not valid and therefore should be removed.
    logger.Warn("Failed to unmarshal pending TLS identity metadata", logger.Ctx{"err": err})
}

    expiredPendingTLSIdentities = append(expiredPendingTLSIdentities, pendingTLSIdentity)
}
```

Also slightly concerned about using `err != nil` as a reason to remove it, as looking at `PendingTLSMetadata` it can return an error for one other reason than invalid JSON.

Reply:
```
// PendingTLSMetadata returns the pending TLS identity metadata.
func (i Identity) PendingTLSMetadata() (*PendingTLSMetadata, error) {
	if i.Type != api.IdentityTypeCertificateClientPending {
		return nil, fmt.Errorf("Cannot extract pending TLS identity secret: Identity is not pending")
	}

	var metadata PendingTLSMetadata
	err := json.Unmarshal([]byte(i.Metadata), &metadata)
	if err != nil {
		return nil, fmt.Errorf("Failed to unmarshal pending TLS identity metadata: %w", err)
	}

	return &metadata, nil
}
```

The one other reason is if identity is not pending, but since we are only querying for pending identities it should not raise any immediate problems. That said, I think it would be nice to check for a specific error here as to not rely so much on another function's specific implementation.

Reply:
I'm confused about this. The only identities that are returned by the query are pending TLS identities. If we can't get their metadata then, regardless of the reason, they are invalid since they cannot be used e.g. when an untrusted used tries to activate it with a token. We should never enter this block but the warning is their in case we do.

I can add a new error type e.g. `ErrInvalidIdentityType` or similar to check if the error was returned by the other stanza but in this case we'd still want to delete it

Reply:
OK as long as an error from this function always means delete the identity token.

But in general we should avoid assuming that all errors from a function mean a single thing, as that can change over time and introduce logic bugs.

It would generally be preferable to be able to identify the specific error class that should lead to the deletion, and anything unexpected is then logged.

Reply:
It always means delete the token as long as we're sure that the identity is a pending identity. The function itself is defined on the general identity... I'm beginning to understand the problem a bit better now.

I think what we can do in this case if have an error type defined in the DB:
```go
type InvalidIdentityMetadataError struct {
    err error
}
```
which is only returned if the JSON unmarshalling fails, but not if e.g. `CertificateMetadata()` is called on an OIDC identity. We should return this error when getting metadata for any identity type. Getting this error will always mean the identity is invalid and must be deleted.

Reply:
I suppose another way of looking at this is that `PendingTLSMetadata()` could just not return an error ever, and if it returns nil metadata (because its the incorrect identity type or invalid metadata) then we can then treat it the same way.

Reply:
Yeah I'm happy to go with that, though I'd like to do the same for all identity types. Can I update this in a separate PR?

Reply:
Though I suppose this might mask the underlying error and make it harder to reason about why the unmarshalling failed. We can add a warning log for it though

Reply:
hrm in that case dont warn in the function but return an error then.

But im confused, because we werent logging the error anyway, just skipping over it.

Reply:
I have updated `PendingTLSMetadata` to return a 400 Bad Request error if it is called against an identity that is not a pending TLS identity, and a 500 Internal Server Error if JSON unmarshalling fails. I've added a comment to the token prune task to explain why we want to remove the pending identity in either case.